### PR TITLE
`to_hetero_with_bases(model)`

### PR DESCRIPTION
**Pull request by rusty1s**
Screen Shot 2021-07-07 at 12 36 08
Screen Shot 2021-07-07 at 12 36 20

### CODE REVIEW

**Comment by yaoyaowd** [comment]
For the `Relu` in the graph, does it mean that we concat the result from 3 different `SageConv` and run `Relu` once? I feel it is more memory efficient and easy to understand when combine `Relu` into `SageConv`.

**Comment by rusty1s** [comment]
> For the `Relu` in the graph, does it mean that we concat the result from 3 different `SageConv` and run `Relu` once? I feel it is more memory efficient and easy to understand when combine `Relu` into `SageConv`, or connect each `SageConv` by an individual `Relu`.

This is not correct by definition. In the basis-decomposition scheme, we operate on a single feature matrix rather than distinict node feature matrices for different node types. The GNN layers will be replicated by the number of bases, whose results are then grouped again. As such, every operation outside of the GNN layer is only applied once.

**Comment by rusty1s** [comment]
Updated the Figure based on @yaoyaowd's suggestions :)

File: torch_geometric/nn/to_hetero_with_bases_transformer.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+        super().__init__(module, input_map, debug)
+        self.metadata = metadata
+        self.num_bases = num_bases
+        self.in_channels = in_channels or {}
+        assert len(metadata) == 2
+        assert len(metadata[0]) > 0 and len(metadata[1]) > 0
+
+        # Compute deterministic IDs for each node and edge type:
+        self.node_type2id = {k: i for i, k in enumerate(metadata[0])}
+        sorted_edge_types = sorted(
```
Comment (yaoyaowd) [inline torch_geometric/nn/to_hetero_with_bases_transformer.py:155]: Is there a reason edge type need to be sorted?

File: torch_geometric/nn/to_hetero_with_bases_transformer.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+            return weight * output
+
+        params = list(module.parameters())
+        device = params[0].device if len(params) > 0 else 'cpu'
+
+        self.convs = torch.nn.ModuleList()
+        for _ in range(num_bases):
+            conv = copy.deepcopy(module)
+            conv.fuse = False
+            conv.basis_weight = Parameter(
```
Comment (yaoyaowd) [inline torch_geometric/nn/to_hetero_with_bases_transformer.py:327]: This part confuse me a little. What is `basis_weight`? Is it defined for every conv?

File: torch_geometric/nn/to_hetero_with_bases_transformer.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+from torch_sparse import SparseTensor
+
+from torch_geometric.nn.fx import Transformer
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.nn.dense import Linear
+
+try:
+    from torch.fx import GraphModule, Graph, Node
+except ImportError:
+    GraphModule = Graph = Node = 'GraphModule', 'Graph', 'Node'
```
[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/nn/to_hetero_with_bases_transformer.py]: Thanks. Fixed it.

File: torch_geometric/nn/to_hetero_with_bases_transformer.py (hunk 2 above)
Comment (rusty1s) [reply torch_geometric/nn/to_hetero_with_bases_transformer.py]: I renamed it to `edge_type_weight` which makes more sense IMO. Here, we register a single scalar weight for each individual edge type, which is used to weight the output message based on edge type. Let me know if this is now more clear.

File: torch_geometric/nn/to_hetero_with_bases_transformer.py (hunk 1 above)
Comment (rusty1s) [reply torch_geometric/nn/to_hetero_with_bases_transformer.py]: Actually no, this is an artifact. I removed it.

**Review by yaoyaowd — APPROVED** [review]

**Pull request merged**